massgov/mayflower
# DP-24286 [a11y] Main nav

Any PRs being created needs a changelog.txt file before being merged into dev. See: [Change Log Instructions](https://github.com/massgov/mayflower/blob/develop/docs/for-developers/changelog-instructions.md)


## Description

- Reorder the components to adjust the navigation order by screen readers.
- Make the top level menu be navigated with arrow keys.
- Add the instructional info. for ESC key use for screen reader users.

## Related Issue / Ticket

- [JIRA issue](https://massgov.atlassian.net/browse/DP-24286)
- [Github issue]()

## Steps to Test


1. [SCREEN READER ONLY] In desktop display, check the markup for the order of the components in the top blue bar.
    This change set the components in the order that makes sense to screen reader users when they navigate the components with arrow keys.  (FYI, Keyboard users don't affected by this change at all as they navigate focusable elements with TAB key. ) In `nav.ma__header__hamburger__nav div.ma__header__hamburger-wrapper`, find its child components in the following order:
	1. `div.ma__header__hamburger__button-container` for the wrapper for the menu button.
	2. `div.ma__header__hamburger__nav-container` for the fly-out menu container.
	3. `div.ma__header__hamburger__utility-nav` for the utility nav container
           
2. Find the menu button has `aria-describedby` to give screen reader users the information for the ESC key use.  The information is presented in `div#menuInfo`.  `div#menuInfo` is not visible to sighted users.

  ```
   <button type="button" aria-expanded="false" aria-label="main menu for mass.gov" aria-describedby="menuInfo" class="ma__header__hamburger__menu-button js-header-menu-button" aria-pressed="false">

   <div id="menuInfo" aria-hidden="true">
      <p>In the menu, hitting escape on top menu items closes the main navigation menu.</p>
</div>
```

3.  [KEYBOARD ONLY] Navigate the main menu with keyboard.

	1. Tab to the menu button. 
	2. Hit ENTER or SPACE to open the fly-out.
	3. Tab to get into the menu container.

4. [KEYBOARD ONLY] When a top menu item with sub menu/content has focus, hit ENTER/SPACE.  (FYI, this affects keyboard users.  Screen readers have their default features assigned to each arrow keys and we don't intend to override them not to confuse screen reader users.)

	1. Its sub menu opens and the first sub menu item has focus.
	2. Navigate the sub menu with RIGHT ARROW.  Focus circulates from top to bottom.  If focus is on the last item, moves focus to the first item.
	3. Navigate the sub menu with LEFT ARROW.  Focus circulates from bottom to top (= backward).  If focus is on the first item, moves focus to the last item.
	4. Hit ESC to leave the sub menu.  Focus is set on the parent top menu.
          
5. [KEYBOARD ONLY] On a top menu item, hit DOWN ARROW.

	1. Its sub menu opens and set focus on the first sub menu.
	2. Hit ESC to leave the sub menu.  Focus is set on the parent top menu.
          
7. [KEYBOARD ONLY] On a top menu item, hit TOP ARROW.

	1. Its sub menu opens and set focus on the last sub menu.
	2. Hit ESC to leave the sub menu.  Focus is set on the parent top menu.



## Screenshots
Use something like [licecap](http://www.cockos.com/licecap/) to capture gifs to demonstrate behaviors.


## Additional Notes:

Anything else to add?

#### Impacted Areas in Application


*

#### @TODO


*

#### Today I learned...
